cpp/src/utilities/column_utils.hpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2019 BlazingDB, Inc.
+ *     Copyright 2019 Eyal Rozenberg <eyalroz@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file Small utility functions for working with @ref gdf_column structures,
+ * which are raw C structures without any guarantees and operators defined for
+ * them, and thus require a bit of care when handling. This care and due
+ * diligence tends to be rather repetitive; hence this file.
+ *
+ * @note since the "right" way to work with these structures is pass references,
+ * but the "C" way of doing the same involves pointers, many functions are
+ * implemented for both.
+ */
+#ifndef UTILITIES_COLUMN_UTILS_HPP_
+#define UTILITIES_COLUMN_UTILS_HPP_
+
+#include <utilities/type_dispatcher.hpp>
+
+#include <cudf.h>
+
+#include <algorithm>
+
+namespace cudf {
+
+template<typename T>
+T* get_data(const gdf_column& column) noexcept { return static_cast<T*>(column.data); }
+
+template<typename T>
+T* get_data(const gdf_column* column) noexcept { return get_data<T>(*column); }
+
+constexpr inline bool is_an_integer(gdf_dtype element_type)
+{
+    return
+        element_type == GDF_INT8  or
+        element_type == GDF_INT16 or
+        element_type == GDF_INT32 or
+        element_type == GDF_INT64;
+}
+
+constexpr inline bool is_integral(const gdf_column& column) noexcept { return is_an_integer(column.dtype); }
+constexpr inline bool is_integral(const gdf_column* column) noexcept{ return is_integral(*column); }
+
+constexpr bool is_nullable(const gdf_column& column) noexcept { return column.valid != nullptr; }
+
+namespace detail {
+
+struct size_of_helper {
+    template <typename T>
+    constexpr int operator()() const noexcept { return sizeof(T); }
+};
+
+} // namespace detail
+
+/**
+ * @brief Returns the size in bytes of values of a column element type.
+ */
+constexpr std::size_t inline size_of(gdf_dtype element_type) {
+    return type_dispatcher(element_type, detail::size_of_helper{});
+}
+
+/**
+ * @brief Returns the size in bytes of each element of a column (a.k.a. the column's width)
+ */
+inline std::size_t width(const gdf_column& col) noexcept { return size_of(col.dtype); }
```

Comment:
The rationale here is that "size of" applies to gdf_dtypes, regardless of column.. It's the equivalent of C++ `sizeof()`. Also, for these scalar types - the meaning of "size" is obvious - the size in memory, in  bytes. For columns, however, the term "size", while idiomatic in C++ to denote the number of elements, is somewhat ambiguous - it could easily be interpreted as the size in bytes of the column memory representation. Now, a column is not a line: Like in an actual table on paper, it has two dimensions - a length and a width. I would actually support  using column _length_ rather than column _size_ everywhere, although that's a matter for a different PR / for the column redesign. The width is the fixed size of the element.

Of course, this metaphor doesn't hold perfectly for string columns, but for those, size_of is imperfect too.

Reply:
You seem to be agreeing that width and size_of do the same thing. Therefore one should be removed.

Reply:
But they don't do the same thing - one acts on columns and the other acts on types. Remember these are utility functions which are by definition non-essential. `is_nullable()` (now `has_validity()`) can always be replaced with a pointer nullptr comparison - but then it "sounds" different when you write the code. I want to be able to forget about the implementation details. Same thing for column widths. I want to be able to say "if these columns have the same width than blah blah", without explicitly referring to their type.

Reply:
Fine. Then back to my original concern: `width(column)` is vague. I think `size_of_dtype(column)` is clear. But that's because it is equally as clear as `size_of(column->dtype)`

I think if you are thinking about the width of the data in a column in bytes, you are OK with thinking about data types.

Reply:
@harrism : I'll drop the `width()` function for now. But I might bring this question up at some point later on, on nv-bar, to see whether people consider this vague or not.

Reply:
Following the discussion on Slack, will change this to `byte_width()` which AFAICT we can all live with. (Still think `width()` is more clear though.)